src/core/common/locator.hpp

```diff
@@ -83,7 +87,11 @@ class InstanceLocator
 #if OPENTHREAD_CONFIG_MULTIPLE_INSTANCE_ENABLE
     Instance &GetInstance(void) const { return *mInstance; }
 #else
+#if !OPENTHREAD_CONFIG_INSTANCE_ON_HEAP
     Instance &GetInstance(void) const { return *reinterpret_cast<Instance *>(&gInstanceRaw); }
+#else
+    Instance &GetInstance(void) const { return *gInstanceRaw; }
```

Comment:
Should able to use same implementation for both case when `gInstanceRaw` is defined to `uint64_t *`.